test/integration/consul-container/libs/agent/container.go

```diff
@@ -0,0 +1,422 @@
+package agent
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"os"
+	"path/filepath"
+	"strconv"
+	"time"
+
+	dockercontainer "github.com/docker/docker/api/types/container"
+	"github.com/docker/docker/pkg/ioutils"
+	"github.com/pkg/errors"
+	"github.com/testcontainers/testcontainers-go"
+	"github.com/testcontainers/testcontainers-go/wait"
+
+	"github.com/hashicorp/consul/api"
+	"github.com/hashicorp/consul/test/integration/consul-container/libs/utils"
+)
+
+const bootLogLine = "Consul agent running"
+const disableRYUKEnv = "TESTCONTAINERS_RYUK_DISABLED"
+
+// consulContainerNode implements the Agent interface by running a Consul agent
+// in a container.
+type consulContainerNode struct {
+	ctx            context.Context
+	client         *api.Client
+	pod            testcontainers.Container
+	container      testcontainers.Container
+	serverMode     bool
+	ip             string
+	port           int
+	datacenter     string
+	config         Config
+	podReq         testcontainers.ContainerRequest
+	consulReq      testcontainers.ContainerRequest
+	certDir        string
+	dataDir        string
+	network        string
+	id             int
+	terminateFuncs []func() error
+}
+
+func (c *consulContainerNode) GetName() string {
+	name, err := c.container.Name(c.ctx)
+	if err != nil {
+		return ""
+	}
+	return name
+}
+
+func (c *consulContainerNode) GetConfig() Config {
+	return c.config
+}
+
+func (c *consulContainerNode) GetDatacenter() string {
+	return c.datacenter
+}
+
+func (c *consulContainerNode) IsServer() bool {
+	return c.serverMode
+}
+
+func startContainer(ctx context.Context, req testcontainers.ContainerRequest) (testcontainers.Container, error) {
+	return testcontainers.GenericContainer(ctx, testcontainers.GenericContainerRequest{
+		ContainerRequest: req,
+		Started:          true,
+	})
+}
+
+// NewConsulContainer starts a Consul agent in a container with the given config.
+func NewConsulContainer(ctx context.Context, config Config, network string, index int) (Agent, error) {
```

Comment:
I use it in two places:
1. For the `hostname` of the container. The big problem I ran into with `auto-encrypt` was that agents need to specify `retry-join` for it to work, meaning the names must be deterministic, but the container names also need to be random to allow parallel runs. My solution was to use random container names, but given them an `agent-X` hostname that would be unique per cluster. This also works because there is a docker network per-cluster that sandboxes the hostnames. I'd like to see this a bit more automated/intuitive, but I was going to leave that to future work.
1. As you said, for the convenience of debugging. Specifically for servers being rotated, it's hard to tell which containers are "new" vs "old", and which ones new agents should be trying to join.